File: 9_排序和搜索/二分/二分答案法/二分加滑窗/378. 有序矩阵中第 K 小的元素.ts
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */
const kthSmallest = function (matrix: number[][], k: number): number {
  const [ROW, COL] = [matrix.length, matrix[0].length]

  let left = matrix[0][0]
  let right = matrix[matrix.length - 1][matrix[0].length - 1] + 1

  while (left <= right) {
    const mid = Math.floor((left + right) / 2)
    if (countNGT(mid) < k) left = mid + 1
    else right = mid - 1
  }

  return left

  // !技巧：左下角开始出发
  function countNGT(mid: number) {
    let [row, col] = [ROW - 1, 0]
    let res = 0

    while (row >= 0 && col < COL) {
      if (matrix[row][col] <= mid) {
        res += row + 1
        col++
      } else {
        row--
      }
    }

    return res
  }
}

export default 1

console.log(
  kthSmallest(
    [
      [1, 5, 9],
      [10, 11, 13],
      [12, 13, 15]
    ],
    8
  )
)

File: 9_排序和搜索/二分/二分答案法/二分加滑窗/378. 有序矩阵中第 K 小的元素.py
# https://leetcode.cn/problems/kth-smallest-element-in-a-sorted-matrix/
# 378. 有序矩阵中第 K 小的元素(1<=k<=n^2)
# -1e9<=nums[i]<=1e9

from bisect import bisect_right
from typing import List


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        def countNGT(mid: int) -> int:
            """有多少个不超过mid的候选"""
            return sum(bisect_right(row, mid) for row in matrix)

        left, right = -int(1e9), int(1e9)
        while left <= right:
            mid = (left + right) // 2
            if countNGT(mid) < k:
                left = mid + 1
            else:
                right = mid - 1
        return left
